### PR TITLE
Add GitVersion alias.

### PR DESCRIPTION
**Pull request by RichiCoder1**
Also reworked process and process arguments.

### CODE REVIEW

**Comment by gep13** [comment]
Woot! It is great to see this getting added! @JakeGinnivan

Am I right in saying that I can call this alias, and get back the json output from GitVersion for later use in my build script?

**Comment by RichiCoder1** [comment]
It parses the JSON output, but yup!

**Comment by gep13** [comment]
Sweet! And this addition was using the latest v3 version of GitVersion?  Is it possible to have cake understand the difference between 2 versions of the same tool?

**Comment by RichiCoder1** [comment]
I just used what was ever latest to be honest. >_>

**Comment by gep13** [comment]
>@RichiCoder1 said...
I just used what was ever latest to be honest. >_>

Makes sense!

@patriksvensson is there a way to have Cake support multiple versions of a tool?  Say the input arguments change, get removed, etc.

**Comment by RichiCoder1** [comment]
You could theoretically have an external package that has something like `Cake.GitVersionV2.dll` `Cake.GitVersionV3.dll`. I don't think there's anything in cake that supports this out of the box.

**Comment by patriksvensson** [comment]
Question is, does it differ that much between versions? In the case of xUnit and xUnit v2 we've added a XUnit2 runner since there is both semantic and breaking changes in their API (and both are widely used), but as long as there's only additions I would say that it's up to the user not to use functionality not supported by the version they're using.

If we really must support different versions (a lot of them), I would actually suggest creating NuGet packages for the deprecated ones.

Whatbdo you think?

**Comment by gep13** [comment]
GitVersion v2 to GitVersion v3 was essentially a complete rewrite, and various new arguments were added onto the command line implementation.  Going forward, there hopefully won't be that many changes, but in this one, there was quite a bit.  I was just curious as to how this was handled.  I think for now, the suggestion would be to upgrade GitVersion to the latest, and not support the older one.

**Comment by patriksvensson** [comment]
The plan from the beginning was to support general tool that's widely used and seldom change such as MSBuild.

To be honest there is no modus operandi in place for how to handle this, but I'm open for suggestions.

**Comment by JakeGinnivan** [comment]
I think not supporting v2 is a good idea. The current plan for v4 is to give the command line interface an overhaul, but that will just be switches not so much functionality. We might be able to have a compat shim which we ship to make it easy for people to transition. If we provide a shim it would just be detecting if the shim is there, then calling it instead. We also might be able to detect legacy command line args and translate transparently.

If we don't add any of those compatibility options then based on the major version you are running against you would use different command line switches. If cake could read the major version of the exe then generate the appropriate command line switches then it could be the same extension.

This is pretty awesome! :+1: We should definitely link and add some notes to the GitVersion docs to make this as discoverable as possible

File: src/Cake.Common/Tools/GitVersion/GitVersionSettings.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        /// <summary>
+        /// Gets or sets the path to the log file.
+        /// </summary>
+        public FilePath LogFilePath { get; set; }
+    }
+
+    /// <summary>
+    /// The git version output type.
+    /// </summary>
+    public enum GitVersionOutput
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/GitVersion/GitVersionSettings.cs]: Move this to a separate file.

File: src/Cake.Common/Tools/NuGet/SetApiKey/NuGetSetApiKey.cs [hunk 2]
```diff
@@ -55,13 +55,15 @@ public void SetApiKey(string apiKey, string source, NuGetSetApiKeySettings setti
...
-            string output = null;
+
             var processSettings = new ProcessSettings
             {
                 Arguments = GetArguments(apiKey, source, settings),
                 RedirectStandardOutput = true
             };
-            Run(settings, null, settings.ToolPath, processSettings, process => output = string.Join("\r\n", process.GetStandardOutput()));
+
+            var process = Run(settings, null, settings.ToolPath, processSettings);
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/NuGet/SetApiKey/NuGetSetApiKey.cs]: Why those changes? Doesn't seem related to GitVersion.

File: src/Cake.Common/Tools/NuGet/SetProxy/NuGetSetProxy.cs [hunk 3]
```diff
@@ -54,13 +54,14 @@ public void SetProxy(string url, string username, string password, NuGetSetProxy
                 throw new ArgumentNullException("settings");
             }
 
-            string output = null;
             var processSettings = new ProcessSettings
             {
                 Arguments = GetArguments(url, username, password, settings),
                 RedirectStandardOutput = true
             };
-            Run(settings, null, settings.ToolPath, processSettings, process => output = string.Join("\r\n", process.GetStandardOutput()));
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/NuGet/SetProxy/NuGetSetProxy.cs]: Why those changes? Doesn't seem related to GitVersion.

File: src/Cake.Common/Tools/NuGet/Sources/NuGetSources.cs [hunk 4]
```diff
@@ -133,9 +133,8 @@ public bool HasSource(string source, NuGetSourcesSettings settings)
                 RedirectStandardOutput = true
             };
 
-            var result = false;
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/NuGet/Sources/NuGetSources.cs]: Why those changes? Doesn't seem related to GitVersion.

**Comment by patriksvensson** [comment]
I think the GitVersion functionality is good, but the PR does too much.

* GitVersion support
* Adds a SwitchArgument
* Returns the process (already solved in #344)
* Defines a new method for `IProcess`.

I think it would be better if the PR were more focused around the GitVersion support and make the other tweaks as separate PR:s (unless I'm mistaken and the functionality is required for the GitVersion tool).

**Comment by patriksvensson** [comment]
I spoke with @RichiCoder1 in the [Gitter room](https://gitter.im/cake-build/cake), and the plan is to postpone merging this PR until #344 has been merged.

File: src/Cake.Common/Tools/GitVersion/GitVersionAliases.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using Cake.Core;
+using Cake.Core.Annotations;
+
+namespace Cake.Common.Tools.GitVersion
+{
+    /// <summary>
+    /// Contains funcionality related to GitVersion.
+    /// </summary>
+    [CakeAliasCategory("GitVersion")]
```
Comment (gep13) [inline src/Cake.Common/Tools/GitVersion/GitVersionAliases.cs]: @RichiCoder1  When you come back around to this, would it be possible to change to to be GitTools, and then add an additional alias category to the GitVersion call below?  Similar to what is done here in the ReSharper tools? 

https://github.com/cake-build/cake/blob/develop/src/Cake.Common/Tools/DupFinder/DupFinderAliases.cs

Eventually, I would like to add aliases for all the GitTools, including GitReleaseManager, GitReleaseNotes, as well as GitVersion.  Assuming that @patriksvensson and @devlead  is ok with that :smile_cat:

**Comment by gep13** [comment]
@patriksvensson  @devlead I have started taking a look at this, and I think I have fixed up all the main concerns that there were in @RichiCoder1 original PR.  This PR now only adds GitVersion alias, and nothing else.

I am nothing into one issue though...

When trying to test with either this:

```
Task("GetVersionInfo")
    .Does(() => 
{
    var result = GitVersion(new GitVersionSettings {
        RepositoryPath = "c:/github/organisations/gittools/gitreleasemanager"
    });
    
    Console.WriteLine(result.ToString());
    Console.ReadLine();
});
```

or this:

```
Task("GetVersionInfo")
    .Does(() => 
{
    var result = GitVersion();
    
    Console.WriteLine(result.ToString());
    Console.ReadLine();
});
```

I am getting the following errors:

image

My eyes are starting to glaze over now, really quite sleepy, but if you can see anything obviously wrong, please let me know and I will fix up.

Plan on looking at this again tomorrow, cheers!

File: src/Cake.Common/Tools/GitVersion/GitVersionAliases.cs (hunk 5 above)
Comment (gep13) [inline src/Cake.Common/Tools/GitVersion/GitVersionAliases.cs]: @devlead with regard to the above comment that I made, I can't actually see how this would work anymore.  I thought I did, but now I am not so sure.  Is there a way to group all these tools under the GitTools alias, but still also group by both GitReleaseManager and GitVersion?

**Comment by patriksvensson** [comment]
@gep13 It looks like the `GitVersion` alias is missing the [CakeMethodAlias](http://cakebuild.net/api/cake.core.annotations/8283ce29) attribute.

If the `GitVersionSettings` isn't in the same namespace as the alias, you can also add a [CakeNamespaceImport](http://cakebuild.net/api/cake.core.annotations/0a570436) attribute to tell Cake what extra namespaces to add. You can see an example of this [here](https://github.com/cake-build/cake/blob/5426a07c672ac2c81534d4cc2883ee7e59e7277f/src/Cake.Common/IO/FileAliases.cs#L34).

**Comment by gep13** [comment]
@patriksvensson that was exactly the problem, thank you! :-)

Did you also see my comment on the line above?

**Comment by gep13** [comment]
@patriksvensson @devlead @RichiCoder1 I "think" this is ready for review as well.  I have done some basic testing, and seems to work to me, interested in your thoughts.  Thanks!

**Comment by patriksvensson** [comment]
@gep13 Not sure about what to do. If there is a need to support multiple categories for an alias on the website, we can add it.

**Comment by gep13** [comment]
What I would love to see would be:

GitTools
  - GitVersion
  - GitReleaseManager
    - Create
    - Publish
    - ...

In the case of ReSharper, the nesting is simply

ReSharper
  - DupFinder
  - InspectCode

since there is no further aliases exposed.  So, you perhaps a change is needed to accommodate this.  It's just an idea, I am happy to leave it as:

GitVersion
GitReleaseManager
  - Create
  - Publish
  - ...

Would have just been nice to show that these some from the same place is all.

**Comment by patriksvensson** [comment]
@gep13 I understand. So three levels instead of two? Might take a little bit of work to get that to work. Maybe we can leave it as it is and revisit this later when we know what we need to do.

**Comment by gep13** [comment]
@patriksvensson oh, absolutely, there is no immediate need for this.  I would see this as a "nice to have" :-)

**Comment by gep13** [comment]
@patriksvensson since this is not actually my PR, I can't change the title of the PR, any chance you can change it?  Thanks!

**Comment by devlead** [comment]
@gep13 what's the status here?

**Comment by gep13** [comment]
@devlead it is next on my list.  The PR was ready to go, but now needs to be refactored to use the new base classes that @patriksvensson created.  I forgot about this one the other day when I was refactoring, as I was filtering open PR's based on who started them, i.e. me, so this one got missed out.  Hopefully get to this today.

**Comment by devlead** [comment]
Excellent @gep13 :+1:

**Comment by gep13** [comment]
@devlead ok, assuming all goes well with the builds, and you are happy, I think this is ready.

File: src/Cake.Common/Tools/GitVersion/GitVersionRunner.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+            if (settings.OutputType != GitVersionOutput.BuildServer)
+            {
+                var jsonString = string.Empty;
+
+                Run(settings, GetArguments(settings), new ProcessSettings { RedirectStandardOutput = true },
+                process => jsonString = string.Join("\n", process.GetStandardOutput()));
+
+                var deserializer = new DataContractJsonSerializer(typeof(GitVersion));
+                return deserializer.ReadObject(new MemoryStream(Encoding.UTF8.GetBytes(jsonString))) as GitVersion;
```
Comment (devlead) [inline src/Cake.Common/Tools/GitVersion/GitVersionRunner.cs]: Any special reason the `MemoryStream` isn't disposed/within a using statement here?

**Comment by devlead** [comment]
@gep13 Looks good, just left one minor comment.

File: src/Cake.Common/Tools/GitVersion/GitVersionRunner.cs (hunk 6 above)
Comment (gep13) [inline src/Cake.Common/Tools/GitVersion/GitVersionRunner.cs]: none that I can think of. This was some of the original code from @RichiCoder1 that I hadn't really looked at.  Will refactor that just now, couple minutes...

Comment (devlead) [inline src/Cake.Common/Tools/GitVersion/GitVersionRunner.cs]: Excellent :+1:

**Comment by gep13** [comment]
@devlead pushed a new version, let me know if this is what you were thinking.

**Comment by devlead** [comment]
@gep13 excellent thanks :+1:
